quarkchain/cluster/shard_state.py

```diff
@@ -1368,33 +1368,32 @@ def get_logs(
             Logger.error_exception()
             return None
 
-    def estimate_gas(self, tx: Transaction, from_address) -> Optional[int]:
+    def estimate_gas(self, tx: Transaction, from_address) -> int:
         """Estimate a tx's gas usage by binary searching."""
         evm_tx_start_gas = tx.code.get_evm_transaction().startgas
         # binary search. similar as in go-ethereum
         lo = 21000 - 1
         hi = evm_tx_start_gas if evm_tx_start_gas > 21000 else self.evm_state.gas_limit
-        cap = hi
 
         def run_tx(gas):
-            evm_state = self.evm_state.ephemeral_clone()  # type: EvmState
-            evm_state.gas_used = 0
-            evm_tx = self.__validate_tx(tx, evm_state, from_address, gas=gas)
-            success, _ = apply_transaction(evm_state, evm_tx, tx_wrapper_hash=bytes(32))
-            return success
+            try:
+                evm_state = self.evm_state.ephemeral_clone()  # type: EvmState
+                evm_state.gas_used = 0
+                evm_tx = self.__validate_tx(tx, evm_state, from_address, gas=gas)
+                success, _ = apply_transaction(
+                    evm_state, evm_tx, tx_wrapper_hash=bytes(32)
+                )
+                return success
+            except Exception:
+                return False
 
         while lo + 1 < hi:
             mid = (lo + hi) // 2
-            try:
-                if run_tx(mid):
-                    hi = mid
-                else:
-                    lo = mid
-            except Exception:
+            if run_tx(mid):
+                hi = mid
+            else:
                 lo = mid
-        if hi == cap and not run_tx(hi):
```

Comment:
this is needed otherwise `run(hi)` won't be tried, if the client specifies the cap.

Reply:
ah I see. will return `hi` anyway

Reply:
we don't need to check the result of `run_tx(hi)`. we have to return `hi` anyway.

Reply:
a second thought, this may not be the best way to handle, say if the client specifies a wrong gas cap (like 21001), the binary search will obviously fail, however the new code will return 21001. I think a better way to handle is return error.

also in the link you provided I don't see returning estimate that will not be enough, it's the other way around, that the estimate could be more than actually needed